src/events/attributes.rs

```diff
@@ -32,6 +35,93 @@ pub struct Attribute<'a> {
+    ///
+    /// This will allocate unless the raw attribute value does not require normalization.
+    ///
+    /// See also [`normalized_value_with_custom_entities()`](#method.normalized_value_with_custom_entities)
+    pub fn normalized_value(&'a self) -> Result<Cow<'a, [u8]>, EscapeError> {
+        self.make_normalized_value(None)
+    }
+
+    /// Returns the attribute value normalized as per the XML specification, using custom entities.
+    ///
+    /// https://www.w3.org/TR/xml/#AVNormalize
+    ///
+    /// Do not use this method with HTML attributes.
+    ///
+    /// Escape sequences such as `&gt;` are replaced with their unescaped equivalents such as `>`
+    /// and the characters \t, \r, \n are replaced with whitespace characters.
+    /// Additional entities can be provided in `custom_entities`.
+    ///
+    /// This will allocate unless the raw attribute value does not require normalization.
+    ///
+    /// See also [`normalized_value()`](#method.normalized_value)
+    ///
+    /// # Pre-condition
+    ///
+    /// The keys and values of `custom_entities`, if any, must be valid UTF-8.
+    pub fn normalized_value_with_custom_entities(
+        &'a self,
+        custom_entities: &CustomEntityMap,
+    ) -> Result<Cow<'a, [u8]>, EscapeError> {
+        self.make_normalized_value(Some(custom_entities))
+    }
+
+    /// Normalize the attribute value according to xml specification section 3.3.3
+    ///
+    /// https://www.w3.org/TR/xml/#AVNormalize
+    fn make_normalized_value(
+        &'a self,
+        custom_entities: Option<&CustomEntityMap>,
+    ) -> Result<Cow<'a, [u8]>, EscapeError> {
+        // TODO: avoid allocation when not needed
+        let mut normalized: Vec<u8> = Vec::with_capacity(self.value.len());
+
+        let attr = self.value.as_ref();
+        let mut attr_iter = attr.iter().enumerate();
+
+        while let Some((idx, ch)) = attr_iter.next() {
+            match ch {
+                b' ' | b'\n' | b'\r' | b'\t' => normalized.push(b' '),
+                b'&' => {
+                    let end = idx
+                        + 1
+                        + attr_iter
+                            .position(|(_, c)| *c == b';')
+                            .ok_or_else(|| EscapeError::UnterminatedEntity(idx..attr.len()))?;
+                    let entity = &attr[idx + 1..end]; // starts after the &
+
+                    if let Some(s) = escapei::named_entity(entity) {
+                        normalized.extend_from_slice(s.as_bytes());
+                    } else if entity.starts_with(b"#") {
+                        let entity = &entity[1..]; // starts after the #
+                        let codepoint = escapei::parse_number(entity, idx..end)?;
+                        escapei::push_utf8(&mut normalized, codepoint);
+                    } else if let Some(value) = custom_entities.and_then(|hm| hm.get(entity)) {
+                        // TODO: recursively apply entity substitution
```

Comment:
@Mingun Does the normal `unescape()` function need to take this into account also?

Reply:
Yes, I think so